allow to create multiple link shares via share api

schiessle opened the pull request:
- [x] Allow to create multiple share links (API level)
- [x] Get multiple share links (API level)
- [x] allow to give share links unique names (API level)
- [x] allow to change unique name of a share (API level)
- [x] Allow to create multiple share links (web ui)
- [x] Show multiple share links (web ui)
- [ ] allow to give share links unique names (web ui)
- [ ] allow to change unique name of a share (web ui)

for ux discussion, etc see  https://github.com/nextcloud/server/issues/11169
Fix #10434
Closes #11537 : not possible, we cannot wait for the request of a new share to be done as the `.done` or `success` hooks generate a new scope and a clipboard triggered by a click needs to be in a direct call of a function triggered by an event :(
Fix #11169

schiessle committed:
allow to create multiple link shares via share api

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
allow to add labels to shares

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle committed:
allow to update lables for public link shares

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle commented:
cc @AndyScherzinger  @tobiasKaminsky @marinofaggiana @camilasan @rullzer  With Nextcloud 15 we will support multiple share links. This PR has already implement everything you need to start implementing it on the client. Just in case you looking for something to do :wink: 

Creating links works the same as before. There is just one additional parameter you can use "label" to create/update link shares so that the user can give the share link names to remember which link was used for what.

marinofaggiana commented:
ToDo added :-)

skjnldsv commented on core/Migrations/Version13000Date20170718121200.php:
```diff
@@ -397,6 +397,10 @@ public function changeSchema(IOutput $output, \Closure $schemaClosure, array $op
 				'notnull' => false,
 				'length' => 64,
 			]);
+			$table->addColumn('label', 'string', [
```
Hum, shouldn't we create a newer file for that?

skjnldsv committed:
Migrate link shares to array

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Copy link outside menu

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fix note display

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
Should be finished today!
@jancborchardt @nextcloud/designers Can we try to find another saying for "Share link" I'm always finding this confusing as it looks like an action and not a statement :)


Can you start a review now?
That would increase my response time to fixes to day for the feature freeze :)

Basically the only thing left is the label handling! How should we do that?
1. Add an edit button?
2. In the popovermenu?

skjnldsv committed:
Fix enforce date

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fix password enforcement

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

jancborchardt commented:
Just beginning to test this! :) First feedback:
- The ➕ icon is not keyboard-accessible, no highlight on focus
- Ideally the whole row should be clickable, not just the ➕
- Not sure it’s screenreader-accessible right now with the "title" on the icon. But it would be when the whole row is the action.

jancborchardt committed:
Use primary color for share link icon background instead of bland grey

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

jancborchardt commented:
Made the background of the share link primary color instead of bland grey. This is in line with the change in the Talk app at https://github.com/nextcloud/spreed/pull/1274

skjnldsv commented:
> Made the background of the share link primary color instead of bland grey. This is in line with the change in the Talk app at [nextcloud/spreed#1274](https://github.com/nextcloud/spreed/pull/1274)

I was afraid it would be too much, but it is spot on!! :+1:

skjnldsv commented:
@nextcloud/designers please review!
What to test:
1. public sharing
2. with expire date enforced
3. with password open by default
4. with password enforced
5. with default expire date 
6. note management

It is still missing the label, but I rather have early feedback to fix things asap!

skjnldsv commented:
> * Ideally the whole row should be clickable, not just the heavy_plus_sign

I'm afraid putting the click to the whole row will also trigger other stuff in the children (popover or such) I thing I will rather keep it on the button only to avoid creating issues :(

rullzer commented:
The hide download checkbox does not work anymore

(is only there for file shares now)

schiessle committed:
adjust unit tests

Signed-off-by: Bjoern Schiessle <bjoern@schiessle.org>

schiessle commented:
> Basically the only thing left is the label handling! How should we do that?
> 1. Add an edit button?
> 2. In the popovermenu?

My suggestion would be to have a "rename" action in the "..." menu. Which behaves similar to renaming a file/folder. If you click the rename action the popover closes and the label "share link" turns into a input field where the user can change the name and press enter or focus out.

rullzer commented on core/js/sharedialoglinkshareview.js:
```diff
@@ -143,46 +158,76 @@
 
 		onHideDownloadChange: function() {
-			var $checkbox = this.$('.hideDownloadCheckbox');
+			var $element = $(event.target);
```
event is not defined here.

rullzer reviewed (approved):
Gave it a test round. Works as advertised.

Lets fix the rest later and get this in so it doesn't grow more in size :wink:

danxuliu committed:
Adjust acceptance tests to changes in link shares

Although now it is possible to create several link shares the acceptance
tests currently handles only the first link share; this first link share
is now created by clicking an "Add new share" button instead of a
checkbox.

Besides that, the "Copy link" button has been moved from the menu to the
row, next to the menu trigger.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

jancborchardt committed:
Fix wording, sort 'Add another link' to bottom so it does not confuse

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

jancborchardt commented:
Pushed some sorting order and wording fixes! Someone needs to build the handlebars stuff though as I am getting errors and there is no docs! (Currently stuck at `Error: Cannot find module 'optimist'`)

----------------------

> My suggestion would be to have a "rename" action in the "..." menu. Which behaves similar to renaming a file/folder. If you click the rename action the popover closes and the label "share link" turns into a input field where the user can change the name and press enter or focus out.

> Yes, I also like this, but the menu is already very filled :(

I would agree with @schiessle here to put it in the  3-dot menu but to be honest this function is just utterly useless.

 -------------------

Because I’m a bit annoyed at this shoveling of stuff into menus and sidebars yet again, I opened an issue for cleaning up that popovermenu: https://github.com/nextcloud/server/issues/12178

Bottom line: We need to stop just stupidly adding stuff! No one will be able to find their favorite custom setting among all the other custom settings. It’s nuts!

danxuliu commented:
>  Someone needs to build the handlebars stuff 

Done.

Besides that I have cleaned up a little the commit history; the code is almost the same as it was before, except for a removed `console.error(e)` (because it was the only content of a commit named _Various fixes_, so I dropped it), an updated documentation to reflect that `ShareItemModel` no longer has the attributes `isLinkShare` and `link`, and a "cast" of the attribute `hideDownload` to a boolean from the int `share.hide_download`.

@skjnldsv If you already made a commit to fix the JavaScript unit tests cherry-picking it on top of the current commits should cause no conflicts.

danxuliu commented on core/js/shareitemmodel.js:
```diff
@@ -864,20 +863,12 @@
 						} else {
 							link += OC.generateUrl('/s/') + share.token;
 						}
```
All the code above to set the link is now dead code, because the `link` attribute is no longer set in the share (and the `url` attribute is now used instead of `link` when copying the link to the clipboard); I do not know what that _pre-token_ thing is, though, and if it can be safely replaced by the URL attribute returned by the server.

skjnldsv replied:
No idea. I usually don't want to touch too much here. 😂

danxuliu reviewed (dismissed):
JavaScript unit tests need to be fixed before merging.

Besides that, adding a new link share after the first one does not work when passwords are enforced (the menu to set the password is not shown).

skjnldsv commented:
> Besides that, adding a new link share after the first one does not work when passwords are enforced (the menu to set the password is not shown).

Nice catch!!

Thanks a lot for the fixes! You two are completely crazy working through the night :(
You need to sleep!

MorrisJobke reviewed (approved):
Tested and works nice 👍 Current state is also perfectly fine to be merged and iterated on later

danxuliu commented:
@jancborchardt By the way, same problem as described in https://github.com/nextcloud/spreed/pull/1274#pullrequestreview-170646080:
![share-link-icon-theming-color-white](https://user-images.githubusercontent.com/26858233/47849559-601a3100-ddd2-11e8-90bb-e58d330e3232.png)

skjnldsv committed:
Password policy fix and menu open on first share

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fixed jsunit

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

danxuliu reviewed (changes requested):
>  Fixed

Not so fast; I have some complaints ;-)

danxuliu commented on core/js/tests/specs/sharedialoglinkshareview.js:
```diff
@@ -81,12 +81,17 @@ describe('OC.Share.ShareDialogLinkShareView', function () {
+					expiration: '',
+					permissions: OC.PERMISSION_READ,
+					stime: 1403884258,
+					id: 123
```
Are all the parameters above needed for the tests? Or is it enough to just give the id?

skjnldsv replied:
No idea if they're needed. The id is at least yes.

danxuliu replied:
No, they were not, so I removed them everywhere ;-)

danxuliu commented on core/js/tests/specs/sharedialoglinkshareview.js:
```diff
@@ -141,16 +151,22 @@ describe('OC.Share.ShareDialogLinkShareView', function () {
 			});
+			view.render();
 
 			$hideDownloadCheckbox = view.$el.find('.hideDownloadCheckbox');
```
No `data-share-id` used here even if the id `321` was used in `shareModel.set` above; maybe it is unneeded in the other test too?

skjnldsv replied:
It is! :)
There is only one share, so it doesn't really matter ;)

danxuliu commented on core/js/tests/specs/sharedialogviewSpec.js:
```diff
@@ -165,47 +177,57 @@ describe('OC.Share.ShareDialogView', function() {
+			shareModel.set({
+				linkShares: [{
+					token: 'thetoken',
+					url: link,
```
The link is now got from the share returned by the server, so it can be hardcoded like the other parameters; no need to generate it above.

skjnldsv commented:
@danxuliu be quick!
I want to get this in today! And don't use this as an excuse to merge your pr before mine :p

skjnldsv commented on core/js/sharedialoglinkshareview.js:
```diff
@@ -537,7 +537,7 @@
 			var minPasswordLength = 4
 			// password policy?
-			if(oc_capabilities.password_policy && oc_capabilities.password_policy.minLength) {
+			if(window.oc_capabilities && oc_capabilities.password_policy && oc_capabilities.password_policy.minLength) {
```
Because it seems there is an issue between the declaration of variables and window.xxx :)

danxuliu replied:
OK, but the regular code is not the place to address that ;-) Values set by [JSConfigHelper](https://github.com/nextcloud/server/blob/master/lib/private/Template/JSConfigHelper.php#L161) should be set in the JavaScript unit tests in [specHelper.js](https://github.com/nextcloud/server/blob/4a0b7aaf6c80f59552b1da861432476a8ac4c7cf/core/js/tests/specHelper.js#L86)

skjnldsv replied:
You learn something everyday!
Well, calling the variable or the window.xxx is the same when declared as global :)

skjnldsv replied:
Done

skjnldsv commented on core/js/tests/specs/sharedialoglinkshareview.js:
```diff
@@ -107,7 +112,6 @@ describe('OC.Share.ShareDialogLinkShareView', function () {
 		it('is not shown if the share is a folder', function() {
 			shareModel.fileInfoModel.set('mimetype', 'httpd/unix-directory');
 
-			// Setting the item type also triggers the rendering
```
No, I changed it because it was unnecessary.
This is done on purpose.

danxuliu replied:
Ha, I can see through your lies! :-P [Setting the item type still triggers the rendering](https://github.com/nextcloud/server/pull/11844/files#diff-da7df34db49581af77bd59e4f269dc36R88) ;-)

Setting the link share does no longer trigger the rendering, though; although you did not that on purpose, it was a side effect of renaming the `linkShare` attribute to `linkShares` in the model :-P (I also [removed the event handler](https://github.com/nextcloud/server/pull/11844/files#diff-da7df34db49581af77bd59e4f269dc36L92) because it was no longer used).

skjnldsv commented on core/js/tests/specs/sharedialoglinkshareview.js:
```diff
@@ -141,16 +151,22 @@ describe('OC.Share.ShareDialogLinkShareView', function () {
 
 			expect($workingIcon.hasClass('hidden')).toBeFalsy();
-			expect(shareModel.saveLinkShare.withArgs({ hideDownload: true }).calledOnce).toBeTruthy();
+			expect(shareModel.saveLinkShare.calledOnce).toBeTruthy();
```
None, forgot to edit ;)

danxuliu reviewed (changes requested):
>  Fixed :D

Almost :-P I will send a final fix.

skjnldsv commented:
Failure unrelated: `sh: 1: kill: No such process` https://drone.nextcloud.com/nextcloud/server/12248/283

skjnldsv commented:
> Almost :-P I will send a final fix.

Ahaha! Okay :)
Thanks a lot for your help @danxuliu :hugs:

danxuliu reviewed (approved):
Rejoice; finally, the nitpicker is _reasonably_ satisfied :+1:

I hope Drone is too... :-P

MorrisJobke commented:
🎉 I will merge once drone has done it's job 🎉

lopezio commented:
> Future interface stuff that is optional:
> > * [ ]  allow to give share links unique names (web ui)
> > * [ ]  allow to change unique name of a share (web ui)

@schiessle @jancborchardt any progress on these points..? I think they're fundamental to make the multiple linking functionality usable..  with some kind of distinction between the shared links it would be possible to explicitely delete or re-share links based on a target public / individual..

Best Regards!
